mobile_packet_verifier/src/burner.rs

```diff
@@ -33,43 +45,130 @@ where
+                        signature = %txn.get_signature(),
+                        %payer,
+                        total_dcs,
+                        max_attempts = self.failed_retry_attempts
+                    );
+
+                    // block on confirmation
+                    self.transaction_confirmation_check(pool, err, txn, payer, total_dcs, sessions)
+                        .instrument(span)
+                        .await;
+                }
             }
+        }
 
-            // We succesfully managed to burn data credits:
+        Ok(())
+    }
 
-            metrics::counter!("burned", "payer" => payer.to_string(), "success" => "true")
-                .increment(total_dcs);
+    async fn transaction_confirmation_check(
+        &self,
+        pool: &Pool<Postgres>,
+        err: S::Error,
+        txn: S::Transaction,
+        payer: PublicKeyBinary,
+        total_dcs: u64,
+        sessions: Vec<pending_burns::DataTransferSession>,
+    ) {
+        tracing::warn!(?err, "starting txn confirmation check");
+        // We don't know if the txn actually made it, maybe it did
 
-            // Delete from the data transfer session and write out to S3
-            pending_burns::delete_for_payer(pool, &payer, total_dcs).await?;
+        let signature = txn.get_signature();
 
-            for session in sessions {
-                self.valid_sessions
-                    .write(ValidDataTransferSession::from(session), &[])
-                    .await?;
+        let mut attempt = 0;
+        while attempt <= self.failed_retry_attempts {
+            tokio::time::sleep(self.failed_check_interval).await;
+            match self.solana.confirm_transaction(signature).await {
+                Ok(true) => {
+                    tracing::debug!("txn confirmed on chain");
+                    let txn_success = handle_transaction_success(
+                        pool,
+                        payer,
+                        total_dcs,
+                        sessions,
+                        &self.valid_sessions,
+                    )
+                    .await;
+                    if let Err(err) = txn_success {
+                        tracing::error!(?err, "txn succeeded, something else failed");
+                    }
+
+                    return;
+                }
+                Ok(false) => {
+                    tracing::info!(attempt, "txn not confirmed, yet...");
+                    attempt += 1;
+                    continue;
+                }
+                Err(err) => {
+                    // Client errors do not count against retry attempts
+                    tracing::error!(?err, attempt, "failed to confirm txn");
+                    continue;
+                }
             }
         }
 
-        Ok(())
+        tracing::warn!("failed to confirm txn");
+
+        // We have failed to burn data credits:
+        metrics::counter!(
+            "burned",
+            "payer" => payer.to_string(),
```

Comment:
do we want to include the payer? this is either just the one service provider or an (potentially) unbounded list of payers that balloons the metrics cost by increasing the breadth of the monitored dataset, right?

Reply:
we already include payer in this metric today.  There could be more wallets in the future that pay for this, but we have some alerts that require payer be in this metric